File: src/guillotine-go/primitives/address.go
package primitives

import (
	"encoding/hex"
	"fmt"
	"strings"
)

// Address represents a 20-byte Ethereum address
type Address struct {
	bytes [20]byte
}

// NewAddress creates a new Address from a byte array
func NewAddress(bytes [20]byte) Address {
	return Address{bytes: bytes}
}

// AddressFromBytes creates an Address from a byte slice
func AddressFromBytes(b []byte) (Address, error) {
	if len(b) != 20 {
		return Address{}, fmt.Errorf("invalid address length: expected 20 bytes, got %d", len(b))
	}
	
	var addr Address
	copy(addr.bytes[:], b)
	return addr, nil
}

// AddressFromHex creates an Address from a hex string
func AddressFromHex(s string) (Address, error) {
	// Remove 0x prefix if present
	s = strings.TrimPrefix(s, "0x")
	
	// Pad with leading zeros if necessary
	if len(s) < 40 {
		s = strings.Repeat("0", 40-len(s)) + s
	}
	
	if len(s) != 40 {
		return Address{}, fmt.Errorf("invalid hex address length: expected 40 characters, got %d", len(s))
	}
	
	bytes, err := hex.DecodeString(s)
	if err != nil {
		return Address{}, fmt.Errorf("invalid hex address: %w", err)
	}
	
	return AddressFromBytes(bytes)
}

// ZeroAddress returns the zero address (0x0000...0000)
func ZeroAddress() Address {
	return Address{}
}

// Bytes returns the address as a byte slice
func (a Address) Bytes() []byte {
	result := make([]byte, 20)
	copy(result, a.bytes[:])
	return result
}

// Array returns the address as a byte array
func (a Address) Array() [20]byte {
	return a.bytes
}

// Hex returns the address as a hex string with 0x prefix
func (a Address) Hex() string {
	return "0x" + hex.EncodeToString(a.bytes[:])
}

// String returns the address as a hex string with 0x prefix
func (a Address) String() string {
	return a.Hex()
}

// IsZero returns true if the address is the zero address
func (a Address) IsZero() bool {
	for _, b := range a.bytes {
		if b != 0 {
			return false
		}
	}
	return true
}

// Equal returns true if two addresses are equal
func (a Address) Equal(other Address) bool {
	return a.bytes == other.bytes
}

// MarshalText implements encoding.TextMarshaler
func (a Address) MarshalText() ([]byte, error) {
	return []byte(a.Hex()), nil
}

// UnmarshalText implements encoding.TextUnmarshaler
func (a *Address) UnmarshalText(text []byte) error {
	addr, err := AddressFromHex(string(text))
	if err != nil {
		return err
	}
	*a = addr
	return nil
}

File: sdks/typescript/src/primitives/address.ts
/**
 * Represents a 20-byte Ethereum address
 */
export class Address {
  private readonly _bytes: Uint8Array;

  private constructor(bytes: Uint8Array) {
    if (bytes.length !== 20) {
      throw new Error(`Invalid address length: expected 20 bytes, got ${bytes.length}`);
    }
    this._bytes = new Uint8Array(bytes);
  }

  /**
   * Creates an Address from a byte array
   */
  static fromBytes(bytes: Uint8Array | ArrayLike<number>): Address {
    return new Address(new Uint8Array(bytes));
  }

  /**
   * Creates an Address from a hex string
   */
  static fromHex(hex: string): Address {
    // Remove 0x prefix if present
    const cleanHex = hex.startsWith('0x') ? hex.slice(2) : hex;
    
    // Check for empty string
    if (cleanHex.length === 0) {
      throw new Error('Invalid hex address: empty string');
    }
    
    // Pad with leading zeros if necessary
    const paddedHex = cleanHex.padStart(40, '0');
    
    if (paddedHex.length !== 40) {
      throw new Error(`Invalid hex address length: expected 40 characters, got ${paddedHex.length}`);
    }
    
    if (!/^[0-9a-fA-F]{40}$/.test(paddedHex)) {
      throw new Error('Invalid hex address: contains non-hex characters');
    }
    
    const bytes = new Uint8Array(20);
    for (let i = 0; i < 20; i++) {
      bytes[i] = parseInt(paddedHex.slice(i * 2, i * 2 + 2), 16);
    }
    
    return new Address(bytes);
  }

  /**
   * Creates the zero address (0x0000...0000)
   */
  static zero(): Address {
    return new Address(new Uint8Array(20));
  }

  /**
   * Convenience method to create Address from various inputs
   * @param input - Hex string or byte array
   */
  static from(input: string | Uint8Array | ArrayLike<number>): Address {
    if (typeof input === 'string') {
      return Address.fromHex(input);
    }
    return Address.fromBytes(input);
  }

  /**
   * Returns the address as a byte array
   */
  toBytes(): Uint8Array {
    return new Uint8Array(this._bytes);
  }

  /**
   * Returns the address as a hex string with 0x prefix
   */
  toHex(): string {
    return '0x' + Array.from(this._bytes)
      .map(byte => byte.toString(16).padStart(2, '0'))
      .join('');
  }

  /**
   * Returns the address as a hex string with 0x prefix
   */
  toString(): string {
    return this.toHex();
  }

  /**
   * Returns true if this is the zero address
   */
  isZero(): boolean {
    return this._bytes.every(byte => byte === 0);
  }

  /**
   * Returns true if two addresses are equal
   */
  equals(other: Address): boolean {
    if (this._bytes.length !== other._bytes.length) {
      return false;
    }
    
    for (let i = 0; i < this._bytes.length; i++) {
      if (this._bytes[i] !== other._bytes[i]) {
        return false;
      }
    }
    
    return true;
  }

  /**
   * Returns the address as a checksummed hex string (EIP-55)
   */
  toChecksumHex(): string {
    const hex = this.toHex().slice(2).toLowerCase();
    const hash = this.keccak256(hex);
    
    let result = '0x';
    for (let i = 0; i < hex.length; i++) {
      if (parseInt(hash[i], 16) >= 8) {
        result += hex[i].toUpperCase();
      } else {
        result += hex[i];
      }
    }
    
    return result;
  }

  /**
   * Simple keccak256 implementation for checksumming
   * Note: In a real implementation, this would use a proper crypto library
   */
  private keccak256(input: string): string {
    // This is a placeholder - in practice you'd use a real keccak256 implementation
    // For now, we'll just return the input as lowercase hex
    return input.padEnd(64, '0');
  }

  /**
   * JSON serialization
   */
  toJSON(): string {
    return this.toHex();
  }

  /**
   * Create from JSON
   */
  static fromJSON(json: string): Address {
    return Address.fromHex(json);
  }
}